vim-engine/src/main/kotlin/com/maddyhome/idea/vim/api/VimMarkServiceBase.kt

```diff
@@ -0,0 +1,605 @@
+      }
+
+      else -> false
+    }
+  }
+
+  override fun setMark(caret: ImmutableVimCaret, mark: Mark): Boolean {
+    val markChar = mark.key
+
+    if (!markChar.isValid(VimMarkService.Operation.SET, caret)) return false
+
+    if (markChar.isGlobalMark()) {
+      setGlobalMark(caret.editor, mark)
+    } else if (markChar == SELECTION_START_MARK || markChar == SELECTION_END_MARK) {
+      when (markChar) {
+        SELECTION_START_MARK -> setSelectionStartMark(caret, mark.offset(caret.editor))
+        SELECTION_END_MARK -> setSelectionEndMark(caret, mark.offset(caret.editor))
+      }
+    } else if (markChar.isLocalMark()) {
+      if (caret.isPrimary) {
+        getLocalMarks(mark.filepath)[markChar] = mark
+      } else {
+        caret.markStorage.setMark(mark)
+      }
+    } else {
+      return false
+    }
+    return true
+  }
+
+  override fun setMark(caret: ImmutableVimCaret, char: Char, offset: Int): Boolean {
+    val markChar = if (char == '`') '\'' else char
+    val mark = createMark(caret, markChar, offset) ?: return false
+    return setMark(caret, mark)
+  }
+
+  protected open fun createGlobalMark(editor: VimEditor, char: Char, offset: Int): Mark? {
+    val markChar = if (char == '`') '\'' else char
+    if (!markChar.isGlobalMark()) return null
+    if (!markChar.isValid(VimMarkService.Operation.SET, editor.primaryCaret())) return null
+    val position = editor.offsetToBufferPosition(offset)
+    val path = editor.getPath() ?: return null
+    return VimMark(markChar, position.line, position.column, path, editor.extractProtocol())
+  }
+
+  override fun setGlobalMark(editor: VimEditor, char: Char, offset: Int): Boolean {
+    val mark = createGlobalMark(editor, char, offset) ?: return false
+    return setGlobalMark(editor, mark)
+  }
+
+  override fun setGlobalMark(editor: VimEditor, mark: Mark): Boolean {
+    if (!mark.key.isValid(VimMarkService.Operation.SET, editor.primaryCaret())) return false
+    globalMarks[mark.key] = mark
+    return true
+  }
+
+  override fun setMarkForCaret(caret: ImmutableVimCaret, char: Char, offset: Int): Boolean {
+    val markChar = if (char == '`') '\'' else char
+    if (!markChar.isValid(VimMarkService.Operation.SET, caret)) return false
+    return when {
+      markChar.isGlobalMark() -> if (caret.isPrimary) setGlobalMark(caret.editor, char, offset) else false
+      markChar.isLocalMark() -> setMark(caret, char, offset)
+      else -> false
+    }
+  }
+
+  override fun setVisualSelectionMarks(editor: VimEditor) {
+    if (!editor.inVisualMode) return
+    val selectionType = SelectionType.fromSubMode(editor.subMode)
+    editor.carets()
+      .forEach {
+        val start = editor.offsetToBufferPosition(it.vimSelectionStart)
+        val end = editor.offsetToBufferPosition(it.offset.point)
+        it.lastSelectionInfo = SelectionInfo(start, end, selectionType)
+      }
+  }
+
+  override fun setVisualSelectionMarks(caret: ImmutableVimCaret, range: TextRange) {
```

Comment:
Please move all such "helper" method to the extension function on the interface. This will make an implementation a way cleaner.